### DISCUSSION TITLE
ref.read inside timer issue

### DISCUSSION THREAD

**Discussion by romaingyh** [rrousselGit/riverpod#2072] [Q&A] [answered]
*English version, French bellow*

Hello, I'm facing an issue when reading/updating a provider inside a Timer. In a large application I have a first page describing a model which lead to a second module with more pages and actions about this model (very simplified but I can give more details if needed)

I want to use a ProviderScope to override the model provider in my second module and not use family modifier in all my providers.
So I define a modelProvider with throws an UnimplementedError by defaults and I override it when I navigate to the screen.
I've a main notifier with a Timer.periodic inside. It's in this Timer I read/update other providers and the issue appear :

> Unhandled Exception: UnimplementedError
> package:riverpod/src/framework/element.dart 679:23  ProviderElementBase.read
> package:flutter_experiments/main.dart 39:22         _TimerNotifier.build.<fn>
> dart:isolate-patch/timer_impl.dart 398:19           _Timer._runTimers
> dart:isolate-patch/timer_impl.dart 429:5            _Timer._handleMessage
> dart:isolate-patch/isolate_patch.dart 192:12        _RawReceivePortImpl._handleMessage
> ===== asynchronous gap ===========================

It seems that my provider isn't override in my timer because of asynchronous character of the timer or something else that I'm missing.

Thanks for your help.

<details>
  <summary>Version française</summary>
  
Bonjour, je mets mon message en français car mon anglais n'est pas très clair je trouve.

Je souhaite passer d'une page A vers B en transmettant un modèle entre les 2. Sur la page B je peux faire beaucoup d'actions sur ce modèle mais je veux éviter de mettre un family modifier dans tous mes providers qui concerne la page B. Je crée donc un modelProvider avec une *UninmplementedError* par défaut puis j'override ce dernier avec mon modèle lorsque que je navigue vers la page B.

Cependant j'ai un notifier lié à la page B avec un timer qui tourne en fond. Dans ce timer je lis d'autres providers et c'est là que l'erreur apparait. J'ai une *UninmplementedError* lorsque je lis le provider qui est override.

Je sens que l'erreur vient de moi, il me manque quelque chose à comprendre mais je n'arrive pas à savoir quoi.

Merci beaucoup pour votre aide.
</details>

Here is a minimal code to reproduce the issue, with the last version of Riverpod (2.1.3) :

<details>
<summary>Code</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const ExperimentsApp());
}

class ExperimentsApp extends StatelessWidget {
  const ExperimentsApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderScope(
      child: MaterialApp(
        title: 'Experiments',
        theme: ThemeData.dark(useMaterial3: true),
        home: const FirstPage(),
      ),
    );
  }
}

final modelProvider = Provider.autoDispose<String>((ref) {
  throw UnimplementedError();
});

final _timerProvider = NotifierProvider<_TimerNotifier, int>(_TimerNotifier.new);

class _TimerNotifier extends Notifier<int> {
  @override
  int build() {
    final timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      state++;

      debugPrint(ref.read(modelProvider).toString());
    });

    ref.onDispose(timer.cancel);

    return 0;
  }
}

class FirstPage extends StatelessWidget {
  const FirstPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('A'),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            Navigator.of(context).push(
              MaterialPageRoute(
                builder: (context) => ProviderScope(
                  overrides: [
                    modelProvider.overrideWithValue('Hello'),
                  ],
                  child: const SecondPage(),
                ),
              ),
            );
          },
          child: const Text('Push'),
        ),
      ),
    );
  }
}

class SecondPage extends ConsumerWidget {
  const SecondPage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('B'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(ref.watch(_timerProvider).toString()),
            Text(ref.watch(modelProvider)),
          ],
        ),
      ),
    );
  }
}
```
</details>

**Comment by rrousselGit** [comment, marked as answer]
Hello!

Your provider throws. Therefore `ref.read` rethrows the error thrown. Remove the `UnimplementedError` from your provider and that should fix the issue

**Reply by romaingyh** [reply]
Thanks for your answer,

Actually I need to throw UninimplementedError until I override my provider. The logic behind is something like this : 

https://codewithandrea.com/articles/flutter-state-management-riverpod/#scoping-providers

**Reply by rrousselGit** [reply]
You didn't override your provider though

**Reply by rrousselGit** [reply]
Well, you did, but deep in the tree. In your case you likely accessed the provider outside of the scope and therefore hit the unoverridden provider

**Reply by rrousselGit** [reply]
Alternatively try specifying `dependencies: [modelProvider]` on your timer provider

**Reply by romaingyh** [reply]
Oh ok that's what I missed, thanks a lot for your quick replies !
